packages/next/build/webpack/plugins/next-esm-plugin.ts

```diff
@@ -254,24 +254,37 @@ export default class NextEsmPlugin implements Plugin {
       }
     }
 
-    compilation.hooks.additionalAssets.tapAsync(
-      PLUGIN_NAME,
-      childProcessDone => {
-        this.updateOptions(childCompiler)
-
-        childCompiler.runAsChild((err, entries, childCompilation) => {
-          if (err) {
-            return childProcessDone(err)
-          }
-
-          if (childCompilation.errors.length > 0) {
-            return childProcessDone(childCompilation.errors[0])
-          }
-
-          this.updateAssets(compilation, childCompilation)
-          childProcessDone()
-        })
-      }
-    )
+    // Hold back the main compilation until our Child Compiler has completed so its assets get optimized
+    let childProcessDone: (err?: Error) => void
+    const child = new Promise(resolve => (childProcessDone = resolve))
+    compilation.hooks.optimizeAssets.tapPromise(PLUGIN_NAME, () => child)
+
+    // Defer the child compiler until known main thread "dead time" (while Terser is doing minification in the background)
+    let started = false
+    compilation.hooks.optimizeChunkAssets.intercept({
+      call: () => {
+        // only run the first time optimizeChunkAssets is called
+        if (started) return
+        started = true
+
+        // Delay the Child Compiler until optimizeChunkAssets has had time to send work to the Terser pool
+        setTimeout(() => {
+          this.updateOptions(childCompiler)
+
+          childCompiler.runAsChild((err, entries, childCompilation) => {
+            if (err) {
+              return childProcessDone(err)
+            }
+
+            if (childCompilation.errors.length > 0) {
+              return childProcessDone(childCompilation.errors[0])
```

Comment:
I think this will mask all the errors from the childCompiler? Should be handled by the `reject` function

Reply:
Whoop! Good catch, I should have checked this when copying over.